org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java

```diff
@@ -394,16 +402,216 @@ public static IClassFile resolveClassFile(URI uri){
 	 * @return either a class file or compilation unit
 	 */
 	public static ITypeRoot resolveTypeRoot(String uriString) {
+		return resolveTypeRoot(uriString, false, null);
+	}
+
+	/**
+	 * Convenience method that combines {@link #resolveClassFile(String)} and
+	 * {@link #resolveCompilationUnit(String)}.
+	 *
+	 * @param uri
+	 * @param returnCompilationUnit
+	 * @param monitor
+	 * @return either a class file or compilation unit
+	 */
+	public static ITypeRoot resolveTypeRoot(String uriString, boolean returnCompilationUnit, IProgressMonitor monitor) {
 		URI uri = toURI(uriString);
 		if (uri == null) {
 			return null;
 		}
 		if (JDT_SCHEME.equals(uri.getScheme())) {
-			return resolveClassFile(uri);
+			IClassFile classFile = resolveClassFile(uri);
+			try {
+				if (returnCompilationUnit && classFile != null && classFile.getSourceRange() == null) {
+					ContentProviderManager contentProvider = JavaLanguageServerPlugin.getContentProviderManager();
+					if (monitor == null) {
+						monitor = new NullProgressMonitor();
+					}
+					String contents;
+					try {
+						contents = contentProvider.getSource(classFile, monitor);
+					} catch (Exception e) {
+						JavaLanguageServerPlugin.logException(e.getMessage(), e);
+						return classFile;
+					}
+					if (contents != null && !contents.isBlank()) {
+						return getWorkingCopy(classFile, contents, monitor);
+					}
+				}
+			} catch (JavaModelException e) {
+				JavaLanguageServerPlugin.logException(e.getMessage(), e);
+			}
+			return classFile;
 		}
 		return resolveCompilationUnit(uri);
 	}
 
+	public static void discardClassFileWorkingCopy(ITypeRoot unit) {

```

Comment:
I don't think unit is ever a ClassFileWorkingCopy for me. Is there a specific case where this should happen ?

Reply:
you are right. Fixed.